### ISSUE TITLE
Compilation broken if CUDA and OpenCL are both there and properly detected

### ISSUE THREAD

**Issue by blattms** [OPM/opm-simulators#3687]
I think my merge of #3619 might have been a premature one.

@akva2 noted that compilation is broken if cuda is installed and opencl bindings are active (at least on Debian bullseye).
```
[ 11%] Building CUDA object CMakeFiles/opmsimulators.dir/opm/simulators/linalg/bda/cusparseSolverBackend.cu.o
/usr/bin/nvcc -forward-unknown-to-host-compiler -ccbin=/usr/bin/c++ -DBOOST_ALL_NO_LIB -DBOOST_ATOMIC_DYN_LINK -DBOOST_DATE_TIME_DYN_LINK -DBOOST_FILESYSTEM_DYN_LINK -DBOOST_SYSTEM_DYN_LINK -DBOOST_THREAD_DYN_LINK -DCL_HPP_MINIMUM_OPENCL_VERSION=120 -DCL_HPP_TARGET_OPENCL_VERSION=120 -DENABLE_TBB=1 -DFMT_HEADER_ONLY -DHAVE_CONFIG_H=1 -DVEXCL_BACKEND_OPENCL -DVEXCL_HAVE_OPENCL_HPP -D_TBB_CPP0X -I/home/mblatt/src/opm/opm-simulators/opm-parallel -I/home/mblatt/src/opm/opm-simulators -I/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi -I/usr/lib/x86_64-linux-gnu/openmpi/include -I/usr/include/suitesparse -I/home/mblatt/src/opm/opm-common -I/usr/include/cjson -I/home/mblatt/src/opm/opm-common/external/fmtlib/include -I/home/mblatt/src/opm/opm-common/opm-parallel/include -I/home/mblatt/src/opm/opm-material -I/home/mblatt/src/opm/opm-grid -I/usr/include/trilinos -I/home/mblatt/src/opm/opm-models -isystem=/home/mblatt/opt/amgcl/include -isystem=/home/mblatt/src/opm/opm-simulators/external/fmtlib/include -isystem=/home/mblatt/opt/vexcl/include -O3 -DNDEBUG -Xcompiler=-fPIC -std=c++17 -x cu -c /home/mblatt/src/opm/opm-simulators/opm/simulators/linalg/bda/cusparseSolverBackend.cu -o CMakeFiles/opmsimulators.dir/opm/simulators/linalg/bda/cusparseSolverBackend.cu.o
/usr/include/CL/opencl.hpp:2122:22: error: 'cl::Device cl::Device::default_' has incomplete type
 2122 |     static Device default_;
      |                      ^~~~~~  
/usr/include/CL/opencl.hpp:2387:23: error: redefinition of 'std::once_flag cl::Device::default_initialized_'
 2387 | CL_HPP_DEFINE_STATIC_MEMBER_ std::once_flag Device::default_initialized_;
      |                       ^~~~~~
/usr/include/CL/opencl.hpp:2121:30: note: 'std::once_flag cl::Device::default_initialized_' previously defined here
 2121 |     static std::once_flag default_initialized_;
      |                              ^~~~~~~~~~~~~~~~~~  
/usr/include/CL/opencl.hpp:2389:15: error: redefinition of 'cl_int cl::Device::default_error_'
 2389 | CL_HPP_DEFINE_STATIC_MEMBER_ cl_int Device::default_error_ = CL_SUCCESS;
      |               ^~~~~~
/usr/include/CL/opencl.hpp:2123:22: note: 'cl_int cl::Device::default_error_' previously declared here
 2123 |     static cl_int default_error_;
      |                      ^~~~~~~~~~~~  
/usr/include/CL/opencl.hpp:2402:24: error: 'cl::Platform cl::Platform::default_' has incomplete type
 2402 |     static Platform default_;
      |                        ^~~~~~  
/usr/include/CL/opencl.hpp:2750:23: error: redefinition of 'std::once_flag cl::Platform::default_initialized_'
 2750 | CL_HPP_DEFINE_STATIC_MEMBER_ std::once_flag Platform::default_initialized_;
      |                       ^~~~~~~~
/usr/include/CL/opencl.hpp:2401:30: note: 'std::once_flag cl::Platform::default_initialized_' previously defined here
 2401 |     static std::once_flag default_initialized_;
      |                              ^~~~~~~~~~~~~~~~~~  
/usr/include/CL/opencl.hpp:2752:15: error: redefinition of 'cl_int cl::Platform::default_error_'
 2752 | CL_HPP_DEFINE_STATIC_MEMBER_ cl_int Platform::default_error_ = CL_SUCCESS;
      |               ^~~~~~~~
/usr/include/CL/opencl.hpp:2403:22: note: 'cl_int cl::Platform::default_error_' previously declared here
 2403 |     static cl_int default_error_;
      |                      ^~~~~~~~~~~~  
/usr/include/CL/opencl.hpp:2785:23: error: 'cl::Context cl::Context::default_' has incomplete type
 2785 |     static Context default_;
      |                       ^~~~~~  
/usr/include/CL/opencl.hpp:3181:23: error: redefinition of 'std::once_flag cl::Context::default_initialized_'
 3181 | CL_HPP_DEFINE_STATIC_MEMBER_ std::once_flag Context::default_initialized_;
      |                       ^~~~~~~
/usr/include/CL/opencl.hpp:2784:30: note: 'std::once_flag cl::Context::default_initialized_' previously defined here
 2784 |     static std::once_flag default_initialized_;
      |                              ^~~~~~~~~~~~~~~~~~  
/usr/include/CL/opencl.hpp:3183:15: error: redefinition of 'cl_int cl::Context::default_error_'
 3183 | CL_HPP_DEFINE_STATIC_MEMBER_ cl_int Context::default_error_ = CL_SUCCESS;
      |               ^~~~~~~
/usr/include/CL/opencl.hpp:2786:22: note: 'cl_int cl::Context::default_error_' previously declared here
 2786 |     static cl_int default_error_;
      |                      ^~~~~~~~~~~~  
/usr/include/CL/opencl.hpp:7083:28: error: 'cl::CommandQueue cl::CommandQueue::default_' has incomplete type
 7083 |     static CommandQueue default_;
      |                            ^~~~~~  
/usr/include/CL/opencl.hpp:8809:23: error: redefinition of 'std::once_flag cl::CommandQueue::default_initialized_'
 8809 | CL_HPP_DEFINE_STATIC_MEMBER_ std::once_flag CommandQueue::default_initialized_;
      |                       ^~~~~~~~~~~~
/usr/include/CL/opencl.hpp:7082:30: note: 'std::once_flag cl::CommandQueue::default_initialized_' previously defined here
 7082 |     static std::once_flag default_initialized_;
      |                              ^~~~~~~~~~~~~~~~~~  
/usr/include/CL/opencl.hpp:8811:15: error: redefinition of 'cl_int cl::CommandQueue::default_error_'
 8811 | CL_HPP_DEFINE_STATIC_MEMBER_ cl_int CommandQueue::default_error_ = CL_SUCCESS;
      |               ^~~~~~~~~~~~
/usr/include/CL/opencl.hpp:7084:22: note: 'cl_int cl::CommandQueue::default_error_' previously declared here
 7084 |     static cl_int default_error_;
      |                      ^~~~~~~~~~~~  
make[2]: *** [CMakeFiles/opmsimulators.dir/build.make:940: CMakeFiles/opmsimulators.dir/opm/simulators/linalg/bda/cusparseSolverBackend.cu.o] Error 1
make[2]: Leaving directory '/home/mblatt/src/opm/opm-simulators/opm-parallel'
make[1]: *** [CMakeFiles/Makefile2:1317: CMakeFiles/opmsimulators.dir/all] Error 2
make[1]: Leaving directory '/home/mblatt/src/opm/opm-simulators/opm-parallel'
make: *** [Makefile:160: all] Error 2
```
The problem might be that nvcc sees the opencl headers when compiling (not 100% sure yet, though):
WellContributions.hpp is included which includes opencl.hpp in this case.

**Comment by blattms** [comment]
@Tongdongq maybe we should do some further refactoring to clearly separate the cuda/opencl functionality?

For WellContributions this might look like this:
- Base Class WellContributions knows nothing about CUDA/OpenCL
- Sublass WellContributionsCuda contains all cuda specific stuff
- Sublcass WellContributionOpenCL contains all OpenCL specific stuff

Alternatively instead of subclasses WellContributionsCuda/WellContributionOpenCL could be classes just using WellContributions (maybe as friends).

Anyway I will test without VexCL/amgCL just to rule them out as causes for this.

**Comment by akva2** [comment]
as an intermediate the opencl data could be put in an opaque struct which is defined in the cpp file.

**Comment by blattms** [comment]
Not sure what you mean by that.

**Comment by akva2** [comment]
in WellContributions.hpp
```
struct OpenCLData;
```
you replace the members at https://github.com/OPM/opm-simulators/blob/master/opm/simulators/linalg/bda/WellContributions.hpp#L94 with such a struct. In the header it's opaque - it's just not defined, it's just some struct (ie a pointer).

then in WellContributions.cpp you have
```
struct OpenCLData {
 cl::Context *context;
    cl::CommandQueue *queue;
    Opm::Accelerator::stdwell_apply_kernel_type *kernel;
    Opm::Accelerator::stdwell_apply_no_reorder_kernel_type *kernel_no_reorder;
    std::vector<cl::Event> events;

    std::unique_ptr<cl::Buffer> d_Cnnzs_ocl, d_Dnnzs_ocl, d_Bnnzs_ocl;
    std::unique_ptr<cl::Buffer> d_Ccols_ocl, d_Bcols_ocl;
    std::unique_ptr<cl::Buffer> d_val_pointers_ocl;

    bool reorder = false;
    int *h_toOrder = nullptr;
};
```
ie, the definition. this way you do not have to pull in opencl.hpp in WellContributions.hpp.

**Comment by Tongdongq** [comment]
I think the opaque pointer approach introduces less new header files.

**Comment by akva2** [comment]
I don't think the amount of header files is a very relevant metric. A refactoring with proper split between different APIs and better encapsulation of those is still warranted in the long run.

**Comment by blattms** [comment]
Fixed in one of the many OpenCL PRs #3690 - #3703.
